File: lib/look.py
#!/usr/bin/env python3
"""Vivid look/curve manifest loader — montaj_assets/luts/looks.json.

The JSON is the canonical source of truth for which color-grade LUTs exist
and which one is the project default ("master look"). No heavy imports here
(JSON + paths only) so this stays cheap to import from anywhere: lib/proxy.py,
lib/normalize.py, cli/, serve/.

Asset path resolution mirrors lib/types/colorspace.py:24 — montaj_assets/ is
a sibling of lib/ at the repo/install root.
"""
import json
from pathlib import Path

_LUTS_DIR = Path(__file__).resolve().parent.parent / "montaj_assets" / "luts"
_MANIFEST_PATH = _LUTS_DIR / "looks.json"
_DATA = json.loads(_MANIFEST_PATH.read_text())

MASTER_LOOK: str = _DATA["masterLook"]
"""The manifest's default look id (e.g. "vivid1") — the curve new projects grade with."""

_CURVES: dict = _DATA["curves"]


def curve_ids() -> list[str]:
    """All registered curve ids, in manifest order."""
    return list(_CURVES.keys())


def lut_path(curve_id: str | None = None) -> Path:
    """Absolute path to the .cube file for `curve_id`.

    `curve_id=None` resolves to MASTER_LOOK. Raises ValueError if `curve_id`
    is not a registered curve.
    """
    if curve_id is None:
        curve_id = MASTER_LOOK
    if curve_id not in _CURVES:
        raise ValueError(
            f"Unknown look curve {curve_id!r}. Expected one of {curve_ids()}. "
            f"Check montaj_assets/luts/looks.json."
        )
    return _LUTS_DIR / _CURVES[curve_id]["file"]

File: montaj_assets/render/look.js
// render/look.js
/**
 * Vivid look/curve manifest — JS loader for montaj_assets/luts/looks.json.
 *
 * Mirrors lib/look.py: the JSON is the canonical source of truth for which
 * color-grade LUTs exist and which one is the project default ("master
 * look"). Path resolution mirrors render/color-space.js (montaj_assets/luts/
 * is a sibling of montaj_assets/render/, same as montaj_assets/schemas/) —
 * cli/commands/install.py's prod-cache copy step copies `luts` alongside
 * `schemas` and `timeline-core` for exactly this reason.
 */
import { readFileSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import { dirname, join } from 'node:path'

const __dirname = dirname(fileURLToPath(import.meta.url))
// LUTs live at montaj_assets/luts/, sibling to this file's parent
// (montaj_assets/render/) — same resolution shape as color-space.js's
// SCHEMA_PATH just above it in that file.
const LUTS_DIR = join(__dirname, '..', 'luts')
const MANIFEST_PATH = join(LUTS_DIR, 'looks.json')

const _DATA = JSON.parse(readFileSync(MANIFEST_PATH, 'utf8'))

/** The manifest's default look id (e.g. "vivid1") — the curve new projects grade with. */
export const MASTER_LOOK = _DATA.masterLook

const _CURVES = _DATA.curves

/** All registered curve ids, in manifest order. */
export function curveIds() {
  return Object.keys(_CURVES)
}

/**
 * Absolute path to the .cube file for `curveId`.
 *
 * `curveId` omitted/null resolves to MASTER_LOOK. Throws if `curveId` is not
 * a registered curve. Mirrors lib/look.py::lut_path.
 */
export function lutPath(curveId = null) {
  const id = curveId ?? MASTER_LOOK
  const curve = _CURVES[id]
  if (!curve) {
    throw new Error(
      `Unknown look curve ${JSON.stringify(id)}. Expected one of `
      + `${JSON.stringify(curveIds())}. Check montaj_assets/luts/looks.json.`
    )
  }
  return join(LUTS_DIR, curve.file)
}
